keps/sig-node/1287-in-place-update-pod-resources/README.md

```diff
@@ -646,6 +668,66 @@ Pod Status in response to user changing the desired resources in Pod Spec.
 * At this time, Vertical Pod Autoscaler should not be used with Horizontal Pod
   Autoscaler on CPU, memory. This enhancement does not change that limitation.
 
+### Lifecycle Nuances
+
+* Terminated containers can be "resized" in that the resize is permitted by the API, and the Kubelet
+  will accept the changes. This makes race conditions where the container terminates around the
+  resize "fail open", and prevents a resize of a terminated container from blocking the resize of a
+  running container (see [Atomic Resizes](#atomic-resizes)).
+* Resizing pods in a graceful shutdown state is permitted.
+
+### Atomic Resizes
+
+A single resize request can change multiple values, including any or all of:
+* Multiple resource types
+* Requests & Limits
+* Multiple containers
+
+These resource requests & limits can have interdependencies that Kubernetes may not be aware of. For
+example, two containers coordinating work may need to be scaled in tandem. It probably doesn't makes
+sense to scale limits independently of requests, and scaling CPU without memory could just waste
+resources. To mitigate these issues and simplify the design, the Kubelet will treat the requests &
+limits for all containers in the spec as a single atomic request, and won't accept any of the
+changes unless all changes can be accepted. If multiple requests mutate the resources spec before
+the Kubelet has accepted any of the changes, it will treat them as a single atomic request.
+
+`AllocatedResources` only accounts for accepted requests, so the Kubelet will need to record
+allocated limits in its internal checkpoint.
+
+Note: If a second infeasible resize is made before the Kubelet allocates the first resize, there can
+be a race condition where the Kubelet may or may not accept the first resize, depending on whether
+it admits the first change before seeing the second. This race condition is accepted as working as
+intended.
+
+### Sidecars
+
+Sidecars, a.k.a. resizeable InitContainers can be resized the same as regular containers. There are
+no special considerations here. Non-restartable InitContainers cannot be resized.
+
+### QOS Class
+
+A pod's QOS class cannot be changed once the pod is started. For in place vertical scaling, this
+means that a pod's resources can be changed to fit with a higher-tier QOS class, but not a lower.
+Even if the resources fit the higher tier, it's QOS class will still remain at the original value.
+
```

Comment:
There is an interesting discussion at https://github.com/kubernetes/kubernetes/pull/123319#discussion_r1514218955 on this_"..But promoting a container from the shared pool to the exclusive pool can be allowed and can be desirable.."_  should we discuss it here in beta or can we take in the future? ping @ffromani

Reply:
Thanks for flagging this. Changing QOS class would require moving cgroups, and I'm not familiar with the nuances of that.

I am trying to reduce the scope for beta as much as possible, in the interest of getting it promoted in v1.32, so if this is something that can be revisited later I'm inclined to go with that direction.

Reply:
I would vote there should be API validations in place on the update request to make sure the container stays in the same QoS class. I think if a user wants to change QoS they should make a new pod. doing the api validation would lower the complexity and different cases IMO

Reply:
I'm proposing API validation to prevent the computed QOS class from being lowered, but I think it's weird to say that for a burstable pod you aren't allowed to set requests == limits, since it would change the QOS class.

Reply:
Are there any special considerations the kubelet makes per QoS class that would cause the results of that to be funky? if I have a suite of best effort pods that are all scaled to practically be guaranteed in all but QoS class, could this confuse e.g the eviction manager? I can't think of any ways but I think we should tie all the loose ends before saying a pod can practically be in a different Qos class in all but cgroup

Reply:
Yes, everything that considers QOS class will need to be updated to use the checkpointed QOS class, rather than the computed QOS class.

Reply:
> Thanks for flagging this. Changing QOS class would require moving cgroups, and I'm not familiar with the nuances of that.
> 
> I am trying to reduce the scope for beta as much as possible, in the interest of getting it promoted in v1.32, so if this is something that can be revisited later I'm inclined to go with that direction.

Absolutely, it can be revisited later. Thanks Tim

Reply:
I've added more details on the implications for all the places QOS class is used. PTAL.

To the above points:
- non-guaranteed pods remain ineligible for exclusive CPU pinning
- eviction is unaffected (looks at usage over requests, doesn't consider QOS class)

Reply:
LGTM

Reply:
> I'm proposing API validation to prevent the computed QOS class from being lowered, but I think it's weird to say that for a burstable pod you aren't allowed to set requests == limits, since it would change the QOS class.

The validation is [already there](https://github.com/kubernetes/kubernetes/blob/af2bf2d8f1b3437fe67e1686266a176320422465/pkg/apis/core/validation/validation.go#L5169-L5171). But +1 for expanding the validation to handle this particular nuance with accompanied clarification in the API docs for [pod QoS class definition](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#quality-of-service-classes).

Reply:
LGTM